### ISSUE TITLE
Is it possible to convert to ONNX format? 

### ISSUE THREAD

**Issue by fettahyildizz** [Nanne/pytorch-NetVlad#88]
Is it possible to convert to ONNX format? We want to work with TensorRT environment for runtime purpose.

**Comment by Nanne** [comment]
I have never looked into this, but would be curious to hear if it works if you give it a try.

**Comment by sonali20** [comment]
@fettahyildizz Did you ever get around to converting this to ONNX format? I need to do that as well

**Comment by fettahyildizz** [comment]
@sonali20 Yes I have converted. It was very long time ago but I believe I have edited source code of the model and made it TorchScriptable. If I find an available time, I would publish a repo to convert Netvlad to ONNX and Tensorrt.

**Comment by sonali20** [comment]
@fettahyildizz Great, thanks. If you are able to share your code, that would be great!

Also, what was your motivation to make it TorchScriptable first? My understanding is that going through TorchScript isn't strictly necessary unless there are control loops but I'm still learning. It seems like the main challenge with NetVLAD is that the clustering info needs to be passed in dynamically but there are no control loops.

**Comment by fettahyildizz** [comment]
@sonali20 I'm working on it, as I said, I don't use it anymore so I need to check my history a bit to share an organized code.

I can't recall my motivation to make it TorchScriptable. My main concern was inference on Tensorrt since I had great runtime limitations.

**Comment by fettahyildizz** [comment]
@sonali20 You are welcomed to use https://github.com/fettahyildizz/netvlad_tensorrt

Please reach to me if you have any problem. I managed to export to ONNX and TRT with this repo.

**Comment by sonali20** [comment]
@fettahyildizz Thank you so much for sharing the repo! All the code looks good but I'm not clear on how you handled the clustering part? From what I understand, the NetVLAD pooling layer needs to be initialized with the cluster centroids for every specific use case - https://github.com/Nanne/pytorch-NetVlad?tab=readme-ov-file#cluster. These centroids have to be computed by running clustering on the features extracted by the encoder (vgg16 or alexnet) from the database images in your specific use case. So this code base enforces a clustering step before inference -

https://github.com/Nanne/pytorch-NetVlad/blob/def998ec105b7acde44c83a40364564c72cc7d72/main.py#L417-L424

I see in your code base that the `netvlad.py` is mostly kept the same as in this one except the`init_params()` function that loads the cluster centroids is removed (and `preprocessImage()` is added) so `self.centroids` will only have the initial random values. What am I missing?

**Comment by fettahyildizz** [comment]
I wanted to inference Netvlad on Tensorrt 8.4.1.5 and I believe this was the only code base I could make it compatible with 8.4.1.5. I'm not familiar with netvlad anymore since it's been more than 2 years but we used Faiss to search between netvlad outputs and it gave great results.

**Comment by sonali20** [comment]
Ah I see. Perhaps that code is just missing then. I see the `use_faiss` flag is set but not actually used

https://github.com/fettahyildizz/netvlad_tensorrt/blob/14a4fd26dfe66e0f2357ba1d94fe30296e08f025/src/netvlad_tensorrt/netvlad.py#L86

**Comment by fettahyildizz** [comment]
@sonali20 Yeah could be, we were querying the most similiar netvlad output using faiss but I removed those parts. I believe that's the reason why that flag is there.

I remember I couldn't find any workaround for Tensorrt 8.4.1.5 with current model so that why I made these changes. Sometimes we added some functionalities to runtime programs when we remove them from models to make it compatible with Tensorrt 8.4.1.5
